datafusion/common/src/logical_type/extension.rs

```diff
@@ -0,0 +1,289 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow_schema::{DataType, FieldRef, IntervalUnit, TimeUnit};
+
+use crate::logical_type::type_signature::TypeSignature;
+use crate::logical_type::LogicalType;
+
+pub type ExtensionTypeRef = Arc<dyn ExtensionType + Send + Sync>;
+
+pub trait ExtensionType: std::fmt::Debug {
+    fn display_name(&self) -> &str;
+    fn type_signature(&self) -> TypeSignature;
+    fn physical_type(&self) -> DataType;
```

Comment:
Would this mean we need to have a variant of each extension type for each physical storage type?

A good built-in example is string. Right now `LogicalType::String` always returns `DataType::String`, but arguably could be `DataType::StringView` or `DataType::LargeString`.

Reply:
Yup! Any type which has different physical representations should probably be implemented through an ExtensionType.